src/app/util/binding-table.cpp

```diff
@@ -20,39 +20,110 @@
  */
 
 #include <app/util/binding-table.h>
-#include <zap-generated/gen_config.h>
 
-static EmberBindingTableEntry bindingTable[EMBER_BINDING_TABLE_SIZE];
+namespace chip {
 
-EmberStatus emberGetBinding(uint8_t index, EmberBindingTableEntry * result)
+BindingTable BindingTable::sInstance;
+
+BindingTable::BindingTable()
+{
+    memset(mNextIndex, EMBER_BINDING_TABLE_SIZE, sizeof(mNextIndex));
+}
+
+CHIP_ERROR BindingTable::Add(const EmberBindingTableEntry & entry)
 {
-    if (index >= EMBER_BINDING_TABLE_SIZE)
+    if (entry.type == EMBER_UNUSED_BINDING)
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+    uint8_t newIndex = GetNextAvaiableIndex();
+    if (newIndex >= EMBER_BINDING_TABLE_SIZE)
+    {
+        return CHIP_ERROR_NO_MEMORY;
+    }
+    mBindingTable[newIndex] = entry;
+    if (mTail == EMBER_BINDING_TABLE_SIZE)
+    {
+        mTail = newIndex;
+        mHead = newIndex;
+    }
+    else
     {
-        return EMBER_BAD_ARGUMENT;
+        mNextIndex[mTail] = newIndex;
+        mTail             = newIndex;
     }
+    mSize++;
+    return CHIP_NO_ERROR;
+}
 
-    *result = bindingTable[index];
-    return EMBER_SUCCESS;
+EmberBindingTableEntry & BindingTable::GetAt(uint8_t index)
+{
+    return mBindingTable[index];
 }
 
-EmberStatus emberSetBinding(uint8_t index, EmberBindingTableEntry * result)
+BindingTable::Iterator BindingTable::RemoveAt(Iterator iter)
 {
-    if (index >= EMBER_BINDING_TABLE_SIZE)
+    if (iter.mTable != this || iter.mIndex == EMBER_BINDING_TABLE_SIZE)
+    {
+        return iter;
+    }
+    if (iter.mIndex == mTail)
     {
-        return EMBER_BAD_ARGUMENT;
+        mTail = iter.mPrevIndex;
     }
+    uint8_t next = mNextIndex[iter.mIndex];
+    if (iter.mIndex != mHead)
+    {
+        mNextIndex[iter.mPrevIndex] = next;
```

Comment:
I don't understand why we need this "build a linked list on top of the array" complexity, and I'm not reviewing its details for now as a result.  Why do we need it?  Why can't we just have the array, entries that are free are marked EMBER_UNUSED_BINDING, remove of iterator marks that one entry unused and advances the iterator's index to the next used entry or end, whichever comes first.